includes/safe-svg-settings.php

```diff
@@ -0,0 +1,103 @@
+<?php
+/**
+ * Safe SVG plugin settings.
+ *
+ * @package safe-svg
+ */
+
+/**
+ * SVG settings class.
+ */
+class safe_svg_settings {
+
+	/**
+	 * Set up the class
+	 */
+	public function __construct() {
+		add_action( 'admin_init', [ $this, 'settings_init' ] );
+		add_filter( 'pre_update_option_safe_svg_upload_roles', [ $this, 'update_capability' ], 10, 2 );
+	}
+
+	/**
+	 * Custom option and settings
+	 */
+	public function settings_init() {
+		register_setting( 'media', 'safe_svg_upload_roles' );
+
+		add_settings_section(
+			'safe_svg_settings',
+			__( 'Safe SVG Settings', 'safe-svg' ),
+			[ $this, 'safe_svg_settings_callback' ],
+			'media'
+		);
+
+		add_settings_field(
+			'safe_svg_roles',
+			__( 'User Roles', 'safe-svg' ),
+			[ $this, 'safe_svg_roles_cb' ],
+			'media',
+			'safe_svg_settings'
+		);
+	}
+
+	/**
+	 * Settings section callback function.
+	 *
+	 * @param array $args The settings array, defining title, id, callback.
+	 */
+	public function safe_svg_settings_callback( $args ) {
+		?>
+		<p id="<?php echo esc_attr( $args['id'] ); ?>">
+			<?php esc_html_e( 'Select user roles who can upload SVG files.', 'safe-svg' ); ?>
+		</p>
+		<?php
+	}
+
+	/**
+	 * User role field callback function.
+	 */
+	public function safe_svg_roles_cb() {
+		$user_roles   = get_editable_roles();
+		$upload_roles = (array) get_option( 'safe_svg_upload_roles', [] );
+
+		foreach ( $user_roles as $role => $info ) :
+			?>
+			<div>
+				<label>
+					<input type="checkbox" name="safe_svg_upload_roles[]" value="<?php echo esc_attr( $role ); ?>" <?php checked( in_array( $role, $upload_roles, true ), true ); ?> /> <?php echo esc_html( $info['name'] ); ?>
```

Comment:
The way this is currently set up, no roles will be selected by default, meaning after this update, site admins will need to go in and select the proper roles before things work again (which I think is going to lead to support requests and complaints).

I would suggest that Admins and probably Editors be selected by default. This makes things a little tricky, as we would need to distinguish between situations where this setting hasn't been saved yet and where someone has purposefully deselected all roles and saved that value.

I _think_ the easiest approach is something like:

1. Modify our `register_setting` call to add a `sanitize_callback` function
2. Within that callback, check that we have an array of data and if so, we should probably do something like `array_map( 'sanitize_text_field', $roles )`. If the data we have isn't in an array (so if no roles are selected, I think the value passed will be `null`) we instead save something like `none`
3. We'll need to modify the `update_capability` method to account for that `none` string, basically a check right at the beginning of that method that returns early if `$new_roles === 'none'`
4. And then instead of just doing `current_user_can( 'safe_svg_upload_svg' )`, we'll need a custom method that checks what `get_option( 'safe_svg_upload_roles' )` returns. If it's empty, we can assume that setting hasn't been initialized yet and we can then check if the current user is an admin or editor to determine if they should have access. If it returns an array value or the value `none`, we can do the normal `current_user_can` check

Reply:
I have added the functionality you specced out here, however, I have made all upload roles selected by default for full backwards compatibility.